### DISCUSSION TITLE
Works on My Sandbox: Verifying xgb_train End-to-End

### DISCUSSION THREAD

**Discussion by paulzim** [michelangelo-ai/michelangelo#1821] [Show and tell]
# Works on My Sandbox: Verifying xgb_train End-to-End

I had a Michelangelo sandbox running on my laptop, and I wanted to poke at a working ML platform once I had one. So I picked an example pipeline from `michelangelo-examples`: [`xgb_train`](https://github.com/michelangelo-ai/michelangelo-examples/tree/main/src/michelangelo_examples/california_housing/pipelines/xgb_train), XGBoost regression on the classic California Housing dataset and ran it end-to-end. That turned into a bit of an adventure. Here's what the pipeline does, what broke, and why "just run it" is an underrated way to spend an afternoon.

## The problem being solved

The data is the "California Housing" dataset, which is ~20,000 homes from the 1990 census, each with room count, building age, area median income, location, and the actual sale price as the target. That target caps at $500,001, presumably so a few Atherton mansions didn't skew the set. In 2026 California, $500,001 gets you a down payment.

It's a stand-in. Any real Michelangelo pipeline with ETAs, fraud risk, whatever, has the same shape (input columns, one number to predict), so it's a small, fast way to prove the *plumbing* works before pointing it at something real.

## What XGBoost is, without the math

XGBoost is a popular algorithm for exactly this "predict a number from some columns" problem: it builds a sequence of simple decision trees ("is median income above $50k? older than 30 years?"), where each tree corrects the mistakes of the ones before it. A few hundred small, weak trees add up to one strong predictor... fast, hard to beat on spreadsheet-shaped data, and everywhere in industry.

## The pipeline, like a recipe

The whole job is four steps, each handing its output to the next:

1. **`feature_prep`** — fetch the raw housing data and split it into a training pile and a validation pile, kept aside so the final accuracy check isn't just the model grading its own homework.
2. **`preprocess`** — reshape the data into the numeric format training expects. Runs on Spark, the big-data engine real-world pipelines use.
3. **`train`** — feed the training pile to XGBoost. Runs on Ray, a framework for spreading compute across machines, mostly here to prove the distributed path works, since this dataset could train on one laptop core in seconds.
4. **`push_step`** — files the finished model and an accuracy report card into Michelangelo's model registry, a catalog others can browse for trained models.

## Under the hood: what each step is actually doing

Each step runs as its own isolated unit of work. The platform spins up dedicated compute (a small Ray or Spark cluster), runs the step, and tears it down. Steps never share a process or in-memory data; each one writes its output to shared storage (MinIO, an S3-compatible store, in my case) and passes along a lightweight *handle* (`DatasetVariable`) pointing at where that data lives. The next step reads the handle and pulls the data itself. That's how Ray and Spark, two unrelated engines, hand work to each other without ever talking directly.

### `feature_prep`

- Runs on a small dedicated Ray cluster (2 head CPUs, 1 worker).
- Prefers a CSV bundled in the package; falls back to scikit-learn's downloader only if that's missing, which is deliberate for sandboxes with restricted internet.
- Loads it into pandas, then `ray.data.from_pandas(...)` into a *Ray Dataset*,  a table Ray can split across machines. Overkill at ~20,000 rows, but the same code path a multi-terabyte dataset would use.
- Keeps the 8 feature columns plus price, and splits off 25% as a validation set the model never trains on, so the final accuracy reflects real generalization, not memorization.
- Never returns data directly.  Each dataset is wrapped in a `DatasetVariable` and written to storage; only the handle moves forward.

### `preprocess`

- Runs on its own Spark cluster (1 driver, 1 executor by default), a different engine, started fresh.
- Reads back the exact data `feature_prep` wrote, now as a *Spark* DataFrame instead of a *Ray* Dataset. Same files, different engine's API.
- The transformation is intentionally trivial: cast every selected column to `float`. A production pipeline built from this template would do real feature engineering here; this step just proves the plug-in point works.
- Writes a fresh pair of train/validation `DatasetVariable`s for `train` to pick up.

### `train`

- Runs on its own Ray cluster sized for real training: 1 head node, 2 workers.
- Pulls the Spark-processed data back into Ray, dropping any leftover `uuid`/`datestr` columns so they're never mistaken for model inputs.
- Checks the cluster's actual CPU count (`ray.cluster_resources()`) before training, reserves half for Ray's own data-shuffling, and splits the rest across workers, so the same code runs unmodified on a 3-core sandbox or a production cluster, no hardcoded worker count.
- Training itself runs once per worker, in parallel: each worker gets only its own slice of rows, converted to an `xgboost.DMatrix` (XGBoost's pre-sorted, pre-bucketed format). `xgboost.train(...)` runs with this pipeline's settings (10 boosting rounds, depth-5 trees); Ray's `XGBoostTrainer` keeps workers coordinated so they jointly build *one* model, exchanging summary stats each round so every worker's trees reflect the whole dataset. That coordination is what "distributed training" means here. A callback checkpoints progress and reports accuracy after each round.
- Returns where the model file landed (`model.ubj`) and the final round's metrics, including the validation RMSE used as this exercise's success signal.

### `push_step`

- Runs on Spark again, for convenience (file I/O, talking to other services) rather than any need for distributed processing.
- The fiddliest part: Ray writes its checkpoint path with no storage prefix (`default/ray_results/run-...`, not `s3://...`), expecting the caller to already know the storage system. This step re-attaches the prefix, then uses `fsspec` to find and download the model file.
- Also reloads the train/validation data as plain pandas tables, archiving exact copies alongside the model for later debugging.
- Saves to MinIO/S3 if remote storage is configured, otherwise local files; pushes to the real model registry over gRPC if given an endpoint (same one `ma model get` reads from), or a fake in-memory one for unit tests.
- Bundles the model (tagged as XGBoost regression), a formal eval report, and both datasets into one request to a shared "push" helper, which uploads everything and reports success/failure per item.

## Why run it inside a fake mini-cluster

Rather than risk a shared or production cluster, this all ran inside `k3d`, a disposable Kubernetes cluster that lives entirely in Docker on a laptop. Same APIs, same building blocks as the real thing, just smaller. It's a safe place to prove a pipeline works before it touches anything that matters.

## Two ways to press "go" (and three places the results can land)

Two ways to kick off the same pipeline:

- **`remote-run`** — a one-off command-line trigger. Fire-and-forget; good for quick testing.
- **`ma pipeline apply` + `dev_run`** — registers the pipeline and the run as trackable records, which is what populates MA Studio's "Pipelines"/"Pipeline Runs" tabs.

Once a run is going, though, there are three separate systems of record — each populated only if something explicitly writes to it:

```mermaid
flowchart LR
    subgraph Triggers["Two ways to press \"go\""]
        A["remote-run<br/>(one-off CLI)"]
        B["ma pipeline apply<br/>+ dev_run"]
    end

    subgraph Layers["Three places results can land"]
        C["Cadence workflow execution<br/>cadence CLI / Web UI"]
        D["Pipeline / PipelineRun catalog<br/>MA Studio: Pipelines, Pipeline Runs"]
        E["Model registry<br/>MA Studio: Trained Models / ma model get"]
    end

    A --> C
    B --> C
    B --> D
    C -->|push_step| E

    classDef trigger fill:#e0f2fe,stroke:#0369a1,color:#0c4a6e;
    classDef layer fill:#fef3c7,stroke:#b45309,color:#78350f;
    class A,B trigger;
    class C,D,E layer;
```

Early on, `remote-run` worked and the model showed up in the registry, but Studio's "Pipelines"/"Pipeline Runs" tabs stayed empty, because `remote-run` talks straight to Cadence and skips the catalog layer entirely. Not a bug, just two separate front doors into the same house (fitting, for a pipeline about housing prices), and only one of them has a mail slot.

<img width="1402" height="571" alt="Screenshot 2026-08-14 at 2 22 47 PM" src="https://github.com/user-attachments/assets/91c1a1dc-412b-4177-be58-f40bc06cb5ce" />

## How I know it actually worked
Two independent signals confirmed success:

- **Accuracy**: training reported a validation RMSE (root mean squared error, which is how far off the guesses were, on average, in the same units as price) of `1.3883880945235523`, matching a known-good baseline exactly.
- **A model landed in the registry**: `ma model get --namespace california-housing` showed the new model, confirming the whole pipeline, not just the training math, completed end to end.

## Why bother running an example end-to-end?

Nothing about this model is meant for anything real. It's 20K rows of 1990 census data. The point was never the model; it was proving the published pipeline actually runs on a fresh setup. Along the way I bumped into some stuff to resolve, like a container `PATH` issue, a missing dataset file, a platform scheduling bug (hiding in the exact `controllermgr` subsystem from [my last post](https://github.com/michelangelo-ai/michelangelo/discussions/1601)), and a resource-sizing issue. None of that would have shown up from reading the code alone. It only surfaced because I actually ran it, start to finish, on a clean checkout.

If you want to poke at it yourself, the pipeline's [here](https://github.com/michelangelo-ai/michelangelo-examples/tree/main/src/michelangelo_examples/california_housing/pipelines/xgb_train), and its [README](https://github.com/michelangelo-ai/michelangelo-examples/blob/main/src/michelangelo_examples/california_housing/pipelines/xgb_train/README.md) has the run instructions.

A couple of questions for the group: What's your favorite "toy" dataset for kicking the tires on a new ML platform: California Housing, Iris, Titanic, something else? And if you've poked at Michelangelo's example pipelines yourself, which one should I run next?
